### PR TITLE
CORE-1693: PUT endpoints for Stream Key and User Key

### PR DESCRIPTION
**Pull request by kare**

### CODE REVIEW

**Comment by fonty1** [comment]
@kare it would really help me to include some documentation on how to interact with these new end points.

**Comment by kare** [comment]
@fonty1 what kind of docs you need?

**Comment by fonty1** [comment]
@kare, an example of the request body for each endpoint and what sort of response code and response body I should expect from a successful PUT would speed up my development/testing.

**Comment by kare** [comment]
@fonty1 is this ok:
```
% curl -v -X PUT -H "Authorization: token tester1-api-key" -H "Content-Type: application/json" -d '{"name":"key name"}' http://localhost:8081/streamr-core/api/v1/streams/ln2g8OKHSdi7BcL-bcnh2g/keys/GTM-dQ2fSfaTq0rUJE8R4wkENpX8Z0S7aOKiRWSJz23A
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8081 (#0)
> PUT /streamr-core/api/v1/streams/ln2g8OKHSdi7BcL-bcnh2g/keys/GTM-dQ2fSfaTq0rUJE8R4wkENpX8Z0S7aOKiRWSJz23A HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: token tester1-api-key
> Content-Type: application/json
> Content-Length: 19
>
* upload completely sent off: 19 out of 19 bytes
< HTTP/1.1 204 No Content
< Server: Apache-Coyote/1.1
< Content-Type: application/json;charset=UTF-8
< Date: Mon, 15 Apr 2019 11:09:53 GMT
<
* Connection #0 to host localhost left intact
```
?

**Comment by fonty1** [comment]
@kare thanks, I can understand that, a few questions:
- Can I have the CURL for the User Key as well?
- I notice that the Stream API Key `permission` is absent from the Stream Key PUT above, is that an optional parameter? User needs to be able to change it. AFAIK the User Key does not have this requirement

Screenshot 2019-04-17 at 14 40 22

File: grails-app/conf/UrlMappings.groovy [hunk 1]
```diff
@@ -52,6 +52,7 @@ class UrlMappings {
 		"/api/v1/streams/$id/publishers"(controller: "streamApi", action: "publishers")
 		"/api/v1/streams/$id/status"(controller: "streamApi", action: "status")
 		"/api/v1/streams/$resourceId/keys"(resources: "keyApi", excludes: ["create", "edit", "update"]) { resourceClass = Stream }
+		"/api/v1/streams/$streamId/keys/$keyId"(method: "PUT", controller: "keyApi", excludes: ["create", "edit", "save", "index"], action: "update") { resourceClass = Stream }
```
Comment (hpihkala) [inline grails-app/conf/UrlMappings.groovy]: `excludes` not needed here. It should only be needed when you use the `resources` argument.

File: grails-app/conf/UrlMappings.groovy [hunk 2]
```diff
@@ -70,6 +71,7 @@ class UrlMappings {
 		"/api/v1/users/me"(method: "DELETE", controller: "userApi", action: "delete")
 
 		"/api/v1/users/me/keys"(resources: "keyApi", excludes: ["create", "edit", "update"]) { resourceClass = SecUser }
+		"/api/v1/users/me/keys/$keyId"(method: "PUT", resources: "keyApi", excludes: ["create", "edit", "index", "save"], action: "update") { resourceClass = SecUser }
```
Comment (hpihkala) [inline grails-app/conf/UrlMappings.groovy]: `excludes` not needed here. It should only be needed when you use the `resources` argument.

File: grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy [hunk 3]
```diff
@@ -123,4 +125,23 @@ class KeyApiController {
...
+		}
+		Map json = new JsonSlurper().parseText((String) request.JSON)
+		if (json.name != null && json.name.trim() != "") {
+			useResource(params.resourceClass, params.resourceId) { res ->
+				k.name = json.name.trim()
+				k.save(flush: true, failOnError: true)
+			}
+		}
+		response.status = 200
+		render([
```
Comment (hpihkala) [inline grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy]: Should we render `k.toMap() as JSON` as the result? (the standard json representation of the `Key` object)

File: grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy [hunk 4]
```diff
@@ -126,22 +127,53 @@ class KeyApiController {
...
+
+	@StreamrApi(authenticationLevel = AuthLevel.USER)
+	def updateStreamKey() {
+		Key k = Key.get(params.keyId)
+		if (k == null) {
+			throw new NotFoundException(k.toString(), params.keyId)
+		}
+		Map json = new JsonSlurper().parseText((String) request.JSON)
+		String permission = request.JSON?.permission
+		if (permission) {
```
Comment (hpihkala) [inline grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy]: Two `if`s can be combined:
`if (permission && permission != "read" && permission != "write") {`

File: grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy [hunk 5]
```diff
@@ -126,22 +127,53 @@ class KeyApiController {
...
+				throw new ApiException(400, "INVALID_PARAMETER", "permission field in json should be 'read' or 'write'.")
+			}
+		}
+		if (json.name != null && json.name.trim() != "") {
+			useResource(Stream, params.streamId) { res ->
+				k.name = json.name.trim()
+				if (permission) {
+					Permission.Operation operation = Permission.Operation.fromString(permission)
+					k.getPermissions().collect().each { p ->
+						k.removeFromPermissions(p)
```
Comment (hpihkala) [inline grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy]: This works but causes unnecessary churn in the database. For example, if the user just updates the key name, the existing permission rows get deleted and new ones created, for no reason. Would be better to first see if any changes need to be done, and then do the minimal changes. Something like (pseudocode):
```
if (operation == Permission.Operation.READ) {
    grant read if it doesn't exist already
    delete write if it exists
} else if (operation == Permission.Operation.WRITE) {
    grant read if it doesn't exist already
    grant write if it doesn't exist already
}
```

File: grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy [hunk 6]
```diff
@@ -126,22 +127,53 @@ class KeyApiController {
...
+					// If granting write, grant also read
+					if (operation == Permission.Operation.WRITE) {
+						permissionService.grant(request.apiUser, res, k, Permission.Operation.READ, false)
+					}
+				}
+				k.save(flush: true, failOnError: true)
+			}
+		}
+		response.status = 200
+		render(k.toMap() as JSON)
```
Comment (hpihkala) [inline grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy]: Echo the permission in the result object, if set in the request

**Comment by kare** [comment]
Tested and they both work.
```
% curl -v -X PUT -H "Authorization: token tester1-api-key" -H "Content-Type: application/json" -d '{"name":"new key name", "permission":"write"}' http://localhost:8081/streamr-core/api/v1/streams/SjiL7mCESUuZftQWTNpaIA/keys/I94sMi5vTBWniXClluDj5ALiC27uPvQDWQegH294Iz3A
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8081 (#0)
> PUT /streamr-core/api/v1/streams/SjiL7mCESUuZftQWTNpaIA/keys/I94sMi5vTBWniXClluDj5ALiC27uPvQDWQegH294Iz3A HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: token tester1-api-key
> Content-Type: application/json
> Content-Length: 45
>
* upload completely sent off: 45 out of 45 bytes
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Mon, 29 Apr 2019 09:08:30 GMT
<
* Connection #0 to host localhost left intact
{"id":"I94sMi5vTBWniXClluDj5ALiC27uPvQDWQegH294Iz3A","name":"new key name","user":null,"permission":"write"}

% curl -v -X PUT -H "Authorization: token tester1-api-key" -H "Content-Type: application/json" -d '{"name":"key name XXXX"}' http://localhost:8081/streamr-core/api/v1/users/me/keys/Bf6UJrHMRmq7lXPZ621otAtNVM02JgQ8WfRwdWW7UIgw
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8081 (#0)
> PUT /streamr-core/api/v1/users/me/keys/Bf6UJrHMRmq7lXPZ621otAtNVM02JgQ8WfRwdWW7UIgw HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: token tester1-api-key
> Content-Type: application/json
> Content-Length: 24
>
* upload completely sent off: 24 out of 24 bytes
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Mon, 29 Apr 2019 09:11:01 GMT
<
* Connection #0 to host localhost left intact
{"id":"Bf6UJrHMRmq7lXPZ621otAtNVM02JgQ8WfRwdWW7UIgw","name":"key name XXXX","user":"tester1@streamr.com"}
```

**Comment by kare** [comment]
@fonty1 Stream Key's don't have an associated user with them, but "API Keys" in "Settings" have.

**Pull request merged**